src/test/perl/path.t

```diff
@@ -290,44 +312,137 @@ rmtree if -d $basetest;
+is($mc->symlink($target_file1, $filelink), CHANGED, "file symlink already exists: nothing done");
+ok($mc->is_symlink($filelink), "$filelink is still a symlink (target changed)");
+is(readlink($filelink), $target_file1, "$filelink has the expected target ($target_file1)");
+my $link_status = $mc->symlink($target_file1, $target_file2);
+ok(! defined($link_status), "symlink failed: existing file not replaced by a symlink");
+is($mc->{fail},
+   "*** cannot symlink $target_file2: it is not an existing symlink",
+   "fail attribute set after symlink failure (existing file not replaced by a symlink)");
+ok($mc->file_exists($target_file2) && ! $mc->is_symlink($target_file2), "File $target_file2 has not be replaced by a symlink");
+$opts{force} = 1;
+is($mc->symlink($target_file1, $target_file2, %opts), CHANGED, "existing file replaced by a symlink (force option set)");
+ok($mc->is_symlink($target_file2), "$target_file2 changed to a symlink");
+is(readlink($target_file2), $target_file1, "$target_file2 has the expected target ($target_file1)");
+my $test_directory = "$basetest/$target_directory";
+$link_status = $mc->symlink($target_file1, "$test_directory", %opts);
+ok (! defined($link_status), "directory not replaced by a symlink (force option set)");
+is($mc->{fail},
+   "*** cannot symlink $test_directory: it is not an existing symlink",
+   "fail attribute set after symlink failure (existing directory not replaced by a symlink)");
+ok($mc->directory_exists($test_directory) && ! $mc->is_symlink($test_directory),
+   "Directory $test_directory has not be replaced by a symlink");
+$opts{check} = 1;
+ok(! $mc->symlink("really_really_missing", $brokenlink, %opts), "broken symlink not created (target existence enforced)");
+ok(! -e $brokenlink, "Broken link has not been created");
+delete $opts{check};
+is($mc->symlink("really_missing", $brokenlink), CHANGED, "broken symlink created");
+ok($mc->is_symlink($brokenlink), "Broken link has been created (target check disabled)");
+
+# noreset=0
+verify_exception("symlink tests", "Failed to create symlink", $symlink_call_count * 2, 0);
+ok($mc->_reset_exception_fail(), "_reset_exception_fail after symlink tests");
+
+# Test hardlink creation
+
+init_exception("hardlink tests");
+
+my $cwd = cwd();
+my $hardlink1 = "$basetest/$target_directory/hardlink1";
+my $hardlink2 = "$basetest/$target_directory/hardlink2";
+# hardlink target must be an absolute path or the link path directory is prepended
+$target_file1 = "$cwd/$basetest/tgtfile1";
+my $target_file3 = "$cwd/$basetest/$target_directory/tgtfile3";
+makefile($target_file1);
+makefile($target_file3);
+is($mc->hardlink($target_file1, $hardlink1), CHANGED, "hardlink created");
+is($mc->hardlink($target_file1, $hardlink1), SUCCESS, "hardlink already exists");
+is($mc->hardlink($target_file3, $hardlink1), CHANGED, "hardlink updated");
+$link_status = $mc->hardlink("missing_file", $hardlink1);
+ok(! $link_status, "hardlink not created (target doesn't exist)");
```

Comment:
`doesn't` -> `has to` (there's no check equivalent as with symlink. it really always has to exist